metricq_wizard_backend/metricq/cluster_scanner.py

```diff
@@ -0,0 +1,541 @@
+                issue_type="missing_metadata",
+                severity="error" if "source" in missing_metadata else "info",
+                source=source,
+                missing_metadata=missing_metadata,
+            )
+        else:
+            await self.delete_issue_report(
+                scope_type="metric",
+                scope=metric,
+                type="missing_metadata",
+            )
+
+    async def check_metric_is_dead(
+        self, client: HistoryClient, metric: str, metadata: JsonDict
+    ) -> None:
+        # TODO tolerance is rather high, because in prod, checks seem to
+        # take a while, which messes up the timings :(
+        allowed_age = Timedelta.from_string("1min")
+        try:
+            rate = metadata["rate"]
+            if not isinstance(rate, (int, float)):
+                logger.error(
+                    "Invalid rate: {} ({}) [{}]",
+                    rate,
+                    type(rate),
+                    metadata,
+                )
+            else:
+                allowed_age += Timedelta.from_s(1 / rate)
+        except KeyError:
+            # Fall back to compute tolerance from interval
+            with suppress(KeyError):
+                interval = metadata["interval"]
+                if isinstance(interval, str):
+                    allowed_age += Timedelta.from_string(interval)
+                elif isinstance(interval, (int, float)):
+                    allowed_age += Timedelta.from_s(interval)
+                else:
+                    logger.error(
+                        "Invalid interval: {} ({}) [{}]",
+                        interval,
+                        type(interval),
+                        metadata,
+                    )
+
+        try:
+            request_time = Timestamp.now()
+            result = await client.history_last_value(metric, timeout=60)
+
+            if result is None:
+                # No data points stored yet. This is bad.
+                await self.create_issue_report(
+                    scope_type="metric",
+                    scope=metric,
+                    issue_type="no_value",
+                    severity="warning",
+                    source=metadata.get("source"),
+                )
+                return
+
+            # we have a valid result, hence, there are stored data points.
+            await self.delete_issue_report(
+                scope_type="metric",
+                scope=metric,
+                issue_type="no_value",
+            )
+
+            age = request_time - result.timestamp
+
+            # Archived metrics are supposed to not receive new data points.
+            # For such metrics, the archived metadata is the ISO8601 string, when
+            # the metric was archived.
+            if age > allowed_age and not metadata.get("archived"):
+                # We haven't received a new data point in a while and the metric
+                # wasn't archived => It's dead, Jimmy.
```

Comment:
```suggestion
                # wasn't archived => It's dead, Jim.
```

https://knowyourmeme.com/memes/hes-dead-jimits-dead-jim